Arm/frost signing

arminsabouri opened the pull request:
Opening fo review but there are still a couple open PRs

TODO
- [x] remove utxos after signings for signer
- [x] create make_tx endpoint 
- [x] unit tests

arminsabouri committed:
feature(btc_server): signing round 1 endpoints

arminsabouri committed:
feature(btc_server): add get signing package cordinator endpoint

arminsabouri committed:
feature(btc_server): get round 2 signature endpoint

arminsabouri committed:
feature(btc_server): send round2 signing package to cordinator

arminsabouri committed:
feature(btc_server): create the aggregate signature from cordinator

arminsabouri committed:
feature(btc_wallet): break out sighash function into own function

arminsabouri committed:
refactor(btc_server): move dkg, cordinator, and signer logic into respective files

arminsabouri committed:
refactor(btc_server): move finalize signing logic to cordinator.rs

arminsabouri committed:
chore(btc_wallet): remove unused `sign_with_tweaked_key` fn

arminsabouri committed:
btc_server: fix lint and cargo fmt

arminsabouri committed:
fix(btc_server): use tx sighash for singing package

arminsabouri committed:
feature(btc_server): add agg signture to psbt input and attempt to finalize

arminsabouri committed:
refactor(btc_wallet): update taproot address creation logic to assume a pre-tweaked key. Tweak step is not being done by the FROST library

arminsabouri committed:
feature(btc_server): passing eth tweak to signing, aggregate and verify FROST methods

arminsabouri committed:
refactor(btc_server): get public key should accept eth tweak as request

arminsabouri committed:
refactor(btc_server): finalize() should return serialized tx with witness data

arminsabouri committed:
feature(btc_server): fix finalize signing

as of this commit the btc_server only signs for eth tweaked single input
txs. This needs to be dynamic and will be built out in future commits

arminsabouri committed:
chore(btc_server): clean up old tests

arminsabouri committed:
chore(btc_server): unit tests and docs for `utils/parse_eth_address()`

arminsabouri committed:
chore(btc_server): update taproot frost library crate

arminsabouri committed:
refactor(btc_server): finalize dkg at the end of round 2

arminsabouri committed:
test(btc_server): break out dkg and signing tests

arminsabouri committed:
feature(btc_server): multi input frost signing

arminsabouri committed:
chore(btc_wallet): remove unused fn params for `gateway_address`

arminsabouri committed:
chore(btc_wallet|btc_server): update eth address propriety fields to be a raw key

arminsabouri committed:
chore(btc_wallet|btc_server): clean up and lint fixes

arminsabouri committed:
docs(btc_server): add rust docs too dkg/signing functions

arminsabouri committed:
feature(btc_server): remove selected inputs and add change utxos after finalize signing from the cordiantor

the same operation will also need to be done on the singers

arminsabouri committed:
tests(btc_server): clean up integration tests

scottmillner commented on bin/btc-server/proto/btc_server.proto:
```diff
@@ -17,6 +15,20 @@ service BtcServer {
+    rpc GetRound1SigningPackage(Round1SigningPackageRequest) returns (Round1SigningPackage) {}
+
+    rpc GetRound2SigningPackage(SignPayload) returns (Round2SigningPackage) {}
+
```
typo: `cordinator` > `coordinator`

scottmillner commented on bin/btc-server/src/cordinator.rs:
```diff
@@ -0,0 +1,302 @@
+use std::collections::BTreeMap;
```
typo: file name `cordinator.rs` > `coordinator.rs`

scottmillner commented on bin/btc-server/src/cordinator.rs:
```diff
@@ -0,0 +1,302 @@
+        self.db.flush()?;
+        Ok((signing_packages, psbt))
+    }
+
```
typo: `retruns` > `returns`

scottmillner commented:
@0xBEEFCAF3
a lot of this logic is outside of my wheelhouse so going through the pr was more of a learning session for me instead of me providing useful feedback

looked great and nothing stood out but again frost is not my area of expertise

awesome effort on this!!

scottmillner reviewed (approved):
🔥 🔥 🔥 🔥 🔥 
that's 5 fires!!

arminsabouri committed:
chore(btc_server): re-generate rpc defs

arminsabouri committed:
tests(btc_server): move integration tests to tests dir

arminsabouri committed:
tests(btc_server): added dkg unit tests

arminsabouri committed:
tests(btc_server): added signing round 1 unit test and refactor frost id derivation with a macro

arminsabouri committed:
refactor(btc_server): create error hierarchy for signing, coordinating and dkg

arminsabouri committed:
feature(btc_server): signers should verify the psbt, utxo and recreate the message before creating partial signatures

arminsabouri committed:
fix(btc_server): fix spending change (no eth tweak) outputs

arminsabouri committed:
feature(btc_server): add util to convert psbt to list of signing packages

arminsabouri committed:
feature(btc_server): add helper method to derive change spk

arminsabouri committed:
refactor(btc_server): remove signing packages sled table and gRPC
interface

Signing packages can be generated from the psbt inputs. Before we would
send each signing package per for each input but this can be
determinsticlly generated by signers and cordinators.

arminsabouri committed:
feature(btc_wallet): define `SIGNING_COMMITMENTS` raw psbt key

arminsabouri committed:
feature(btc_server): add get_psbt endpoint to construct a tx outside the context of any current FROST signing session

arminsabouri committed:
feature(btc_server): add utils to add signing commits and partial sigs to psbt

arminsabouri committed:
feature(btc_server): add utils retireve signing commits and partial sigs from psbt

BatiGencho commented on bin/btc-server/src/cordinator.rs:
```diff
@@ -0,0 +1,334 @@
+        frost_id: frost::Identifier,
+        signing_commitments: Vec<frost::round1::SigningCommitments>,
+    ) -> Result<(), CoordinatorError> {
+        self.db.get_key_package()?.ok_or(CoordinatorError::MissingKeyPackage)?;
```
I believe you should do let _ = self.db.get_key_package()?.ok_or(CoordinatorError::MissingKeyPackage)?; as the get key package returns something. I you do cargo clippy it will complain about an ignored result atm.

BatiGencho commented on bin/btc-server/src/cordinator.rs:
```diff
@@ -0,0 +1,334 @@
+        frost_id: frost::Identifier,
+        partial_sigs: Vec<frost::round2::SignatureShare>,
+    ) -> Result<(), CoordinatorError> {
+        self.db.get_key_package()?.ok_or(CoordinatorError::MissingKeyPackage)?;
```
ditto

BatiGencho commented on bin/btc-server/src/cordinator.rs:
```diff
@@ -0,0 +1,334 @@
+        eth_tweak: &[u8; 20],
+    ) -> Result<frost::VerifyingKey, CoordinatorError> {
+        // try to get pk package from db incase we already did dkg round 3
+        if let Some(pk_package) = self.db.get_public_key_package()? {
+            return Ok(pk_package
+                .verifying_key()
+                .to_owned()
+                .get_tweaked(Some(eth_tweak.as_slice())));
+        }
```
I believe this whole thing could be written in a more compact way as:

return self.db.get_public_key_package()?.map(|pk_package| {

pk_package
                .verifying_key()
                .to_owned()
                .get_tweaked(Some(eth_tweak.as_slice())))
})


I have not tried it, but it seems like it could be compacted to the format above.

BatiGencho commented on bin/btc-server/src/main.rs:
```diff
@@ -423,52 +361,38 @@ impl rpc::BtcServer for App {
+            error!("Failed to get round2 dkg package: {}", e);
+            internal!("Failed to get round2 dkg package: {}", e)
+        })?;
+        let json = serde_json::to_string(&round2_packages).unwrap();
```
Looking at serde_json, I feel this ser is pretty slow compared to many others as you can see in the list here:
https://github.com/djkoloski/rust_serialization_benchmark?tab=readme-ov-file

I have played with bitcode/bincode/postcard/borsh and they are all quite stable and fast. If we need speed here ( :??? ) we could swithc to some of them. Wdyt ?

BatiGencho commented on bin/btc-server/src/main.rs:
```diff
@@ -527,38 +520,29 @@ async fn main() -> Result<(), Box<dyn std::error::Error>> {
 
     let min_signers = 2;
-    let max_signers = 2;
+    let max_signers = 3;
```
I believe this should now become an input arg for the btc server. We need the same for the reth side.

arminsabouri replied:
yes correct! I'll add that

BatiGencho commented:
Great MR! It needs a rebase since I merged the DKG stuff this morning, but if you could do that I believe its good to go. I left a few minor stylistic comments, but feel free to ignore them as none of them is reg. the context of the signing. I will study the impl. today and as soon as you merge it later on, I will crate a new branch and start my signing impl.

arminsabouri committed:
refactor(btc_server): signers will put signing commitments and partial
signatures into the psbt

Previously these values were passed around in the gRPC interface. But
there is no reason why they can't call be in the PSBT under unique keys.
This way the API abstracts away all the frost types and complexities.
The only thing passed around is a PSBT

arminsabouri committed:
chore(btc_server): remove unused round1|2 signing packge db structures

arminsabouri committed:
chore(btc_server): remove unused `GetRound1DkgPackages()` RPC method

arminsabouri committed:
chore: pull all frost-tr crates from the same source

arminsabouri committed:
feature(btc_server): max and min signer should be cli flags

BatiGencho committed:
added grpc handlers, configs and utils

arminsabouri commented:
Going to merge and address any existing or additional feedback in future PRs
